src/Expr.cc

```diff
@@ -1429,8 +1429,13 @@ AddToExpr::AddToExpr(ExprPtr arg_op1, ExprPtr arg_op2)
         // We need the IsVector(bt2) check in the following because
         // same_type() always treats "any" types as "same".
         if ( IsVector(bt2) && same_type(t1, t2) ) {
-            SetType(t1);
-            return;
+            // If the RHS is an unspecified vector and the LHS is a
+            // vector-of-vector-of... then treat this as adding an
+            // element, not as an append.
+            if ( t1->Yield()->Tag() != TYPE_VECTOR || ! t2->AsVectorType()->IsUnspecifiedVector() ) {
```

Comment:
```
global vv: vector of vector of count;

global empty: vector of count = vector();

vv += vector(vector(1), vector(2));
vv += vector(vector());
vv += vector(3);
vv += vector();
vv += vector(4);
vv += empty;
vv += vector(5);
print "vv", vv;
```
With 6.1:
```
$ zeek test.zeek 
vv, [[1], [2], [], [3], [4], [], [5]]
```

With this change:
```
$ zeek test.zeek 
vv, [[1], [2], [], [3], [], [4], [], [5]]
```

That makes sense.

I find the conditions, nesting and comments pretty hard to grasp, `! is unspecified` also triggers double negation for me :sweat_smile:  Maybe a helper to make this higher-level?
```
 
+static bool want_element_wise_vector_append(const TypePtr& lhs, const TypePtr& rhs) {
+    if ( ! (IsVector(lhs->Tag()) && IsVector(rhs->Tag())) )
+        return false;
+
+    if ( ! same_type(lhs, rhs) )
+        return false;
+
+    // LHS is not a vector-of-vector, do element-wise-append.
+    if ( lhs->Yield()->Tag() != TYPE_VECTOR )
+        return true;
+
+    if ( rhs->AsVectorType()->IsUnspecifiedVector() )
+        // This is a vector-of-vector-of-X += vector() construct.
+        // It is *not* treated as element-wise-append of an
+        // empty RHS, instead append an empty vector to the LHS.
+        return false;
+
+    return true;
+}
+
 AddToExpr::AddToExpr(ExprPtr arg_op1, ExprPtr arg_op2)
     : BinaryExpr(EXPR_ADD_TO, std::move(arg_op1), std::move(arg_op2)) {
     if ( IsError() )
@@ -1426,18 +1446,14 @@ AddToExpr::AddToExpr(ExprPtr arg_op1, ExprPtr arg_op2)
     }
 
     else if ( IsVector(bt1) ) {
-        // We need the IsVector(bt2) check in the following because
-        // same_type() always treats "any" types as "same".
-        if ( IsVector(bt2) && same_type(t1, t2) ) {
-            // If the RHS is an unspecified vector and the LHS is a
-            // vector-of-vector-of... then treat this as adding an
-            // element, not as an append.
-            if ( t1->Yield()->Tag() != TYPE_VECTOR || ! t2->AsVectorType()->IsUnspecifiedVector() ) {
-                SetType(t1);
-                return;
-            }
+        // Treat += of two vectors as appending each element
+        // of the RHS to the LHS if types agree.
+        if ( want_element_wise_vector_append(t1, t2) ) {
+            SetType(t1);
+            return;
         }
 
+        // Append RHS as element to LHS.
         is_vector_elem_append = true;
 
         bt1 = t1->AsVectorType()->Yield()->Tag();

```

Reply:
Good thought! Now pushed.